File: c/7.c
// What is the 10 001st prime number?

#include <stdio.h>

int divisible(int val, int nums[], int len);

int main()
{
    int iter = 12; // `0 % x` is a "floating point exception"
    int idx = 5; // we have primes at 0..4 already

    // only looking for 10,001 primes
    // and we have these to start with
    int arr[10001] = {2, 3, 5, 7, 11}; 

    while (idx < 10001) // zero is falsy
    {
        if(divisible(iter, arr, idx))
        {
            arr[idx] = iter;
            ++idx;
        }
        ++iter;
    }   

    printf("%d\n", arr[10000]); // 104743

}

int divisible(int val, int nums[], int len)
{
    for (int i = 0; i < len; ++i)
    {
        if (!(val % nums[i]))
            return 0; 
    }
    return 1;
}

File: js/7.js
// What is the 10 001st prime number? <<<--->>>

function prime_test(n, arr) {
    for (var i = 0; i < arr.length; i++) {
        if (n % arr[i] == 0) {
            return false;
        }
    }
    return true;
}

function eratosthenes(n) {
    var primes = [2, 3, 5, 7, 11, 13]; // as listed by projecteuler
    var i = primes[primes.length-1] + 1; // start at next candidate
    while (primes.length < n) {
        if (prime_test(i, primes)) {
            primes.push(i);
        }
        i++;
    }
    return primes;
}

var sieved = eratosthenes(10001);

console.log( sieved[sieved.length-1] ); // 104743   